### PR TITLE
[TF] Moved most Tensor APIs to 'tensorflow/swift-apis'.

### PR DESCRIPTION
**Pull request by eaplatanios**
This PR is moving stuff over from stdlib to swift-apis and switching to using `eager` mode raw op bindings.

I have also incorporated the following updates to tensorflow/swift-apis#109:
- PR #24249

Friend PRs: tensorflow/swift-apis#109 and tensorflow/swift-bindings#26 .

### CODE REVIEW

**Comment by rxwei** [comment]
Since we are preparing v0.3, we'll get to this next week!

**Comment by eaplatanios** [comment]
Sounds good! I'll keep making changes to this PR and the swift-apis one until then, if that's ok.

**Comment by rxwei** [comment]
Sure thing.

**Comment by rxwei** [comment]
Could you resolve the conflicts?

**Comment by eaplatanios** [comment]
@rxwei The conflicts should be resolved now.

File: test/TensorFlowRuntime/tensor_api.swift [hunk 1]
```diff
@@ -47,4 +44,4 @@ TensorNonTPUTests.testAllBackends("BroadcastTensor") {
   expectEqual(Tensor(repeating: 1, shape: [2, 3, 4]), target)
 }
 
-runAllTests()
+runAllTests()
```
Comment (rxwei) [inline test/TensorFlowRuntime/tensor_api.swift]: Could you revert changes to this file?

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow Linux

File: test/TensorFlowRuntime/tensor_api.swift (hunk 1 above)
Comment (eaplatanios) [reply test/TensorFlowRuntime/tensor_api.swift]: Sorry I thought I had. It's reverted now. It was only missing some newlines so it shouldn't affect the tests.

**Comment by rxwei** [comment]
Great, I'm waiting for initial test results on [Linux + CPU](https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/2108/) and [Linux + GPU](https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux-gpu/1295/)

**Comment by eaplatanios** [comment]
Great, thanks Richard! :)

**Comment by rxwei** [comment]
Ok, all tests are passing on CPU. The GPU issues are still the ones introduced in the advanced indexing PR. When you have a chance, could you take a look at it?

File: stdlib/public/TensorFlow/ArrayOps.swift [hunk 2]
```diff
@@ -196,4 +196,43 @@ public extension Raw {
     }
     return out
   }
+
+  /// Unpacks a tensor into `numSplit` tensors along one dimension.
```
Comment (rxwei) [inline stdlib/public/TensorFlow/ArrayOps.swift]: Look like there's no parameter called `numSplit`. Is it the one called `num`?

```suggestion
  /// Unpacks a tensor into `num` tensors along one dimension.
```

**Comment by eaplatanios** [comment]
I can try to look into later today but it may take a while because I need to recompile S4TF on one of our GPU servers. Also, I noticed from the tests that they only fail when we assign to a tensor, not when we're simply indexing. I'll try to figure out why.

**Comment by eaplatanios** [comment]
@rxwei also, I'm not sure if you're aware of this, but when I try to use `#tfop` from outside of stdlib, I get linkage errors for symbols such as `_TFE_DeleteOp`, `_TFE_NewOp`, `_TFE_OpSetAttrInt`, etc.

**Comment by rxwei** [comment]
> @rxwei also, I'm not sure if you're aware of this, but when I try to use `#tfop` from outside of stdlib, I get linkage errors for symbols such as `_TFE_DeleteOp`, `_TFE_NewOp`, `_TFE_OpSetAttrInt`, etc.

Yeah, this is a known bug.

**Comment by eaplatanios** [comment]
Why can't we just export these symbols when compiling `CTensorFlow`?

**Comment by rxwei** [comment]
It's a linker flag issue. Symbols should already be exported module-wise.

**Comment by eaplatanios** [comment]
Where would be a good place to look into for fixing that?

**Comment by rxwei** [comment]
> Where would be a good place to look into for fixing that?

I think that @dan-zheng knows this best!

**Comment by dan-zheng** [comment]
> @rxwei also, I'm not sure if you're aware of this, but when I try to use `#tfop` from outside of stdlib, I get linkage errors for symbols such as `_TFE_DeleteOp`, `_TFE_NewOp`, `_TFE_OpSetAttrInt`, etc.

Could you please share a reproducer?

---

Here's some quick experimentation:
```swift
// tfop.swift
import TensorFlow

let x = Tensor<Float>(1)
let tensor: Tensor<Float> =
  #tfop("Identity", x, T$dtype: Float.tensorFlowDataType,
        __shapes: [x.shape])
print(tensor)
```

```console
$ swift tfop.swift
1.0

$ swiftc tfop.swift
Undefined symbols for architecture x86_64:
  "_TFE_DeleteOp", referenced from:
      _$s10TensorFlow0A0V4tfopE10toHostTest5shapeACyxGAA0A5ShapeV_tF in tfop-5805d4.o
  "_TFE_NewOp", referenced from:
      _$s10TensorFlow0A0V4tfopE10toHostTest5shapeACyxGAA0A5ShapeV_tF in tfop-5805d4.o
  "_TFE_OpSetAttrType", referenced from:
      _$s10TensorFlow0A0V4tfopE10toHostTest5shapeACyxGAA0A5ShapeV_tF in tfop-5805d4.o
  "_TF_DeleteStatus", referenced from:
      _$s10TensorFlow0A0V4tfopE10toHostTest5shapeACyxGAA0A5ShapeV_tF in tfop-5805d4.o
  "_TF_NewStatus", referenced from:
      _$s10TensorFlow0A0V4tfopE10toHostTest5shapeACyxGAA0A5ShapeV_tF in tfop-5805d4.o
ld: symbol(s) not found for architecture x86_64
<unknown>:0: error: link command failed with exit code 1 (use -v to see invocation)

$ swiftc -ltensorflow tfop.swift # succeeds

$ ./tfop
1.0
```

**Comment by eaplatanios** [comment]
Thanks @dan-zheng ! I actually wasn't using `-ltensorflow` because I assumed it would be included by default in S4TF. It's a bit surprising that it is needed for this but not for using other functions provided by the `TensorFlow` library more generally. I also noticed that, irrespective of `#tfop`, if I have a SwiftPM package that compiles fine using `swift build`, linking fails when I use `swift build -c release`. This is resolved by using `swift build -c release -Xlinker -ltensorflow` instead. However, I find this behavior a bit surprising.

**Comment by eaplatanios** [comment]
I'm closing this in favor of a new PR I'm about to open.

**Pull request closed without merging**